### PR TITLE
BUG: use uuid4 instead of md5 for fake column in cross merge

### LINKED ISSUE #48024: BUG: Merge "cross" will assert on systems with FIPS enforcement
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import ssl
import pandas as pd

# if you have a FIPS capable SSL library
ssl.FIPS_mode_set(1)

df = pd.merge(
  pd.DataFrame([...]),
  pd.DataFrame([...]),
  how='cross'
)
```

### Issue Description

When performing a cross merge on a system with [FIPS Mode](https://www.nist.gov/standardsgov/compliance-faqs-federal-information-processing-standards-fips) enabled, it will raise a ValueError when attempting to use a non-secure hashing algorithm (i.e. md5):
```
>       cross_col = f"_cross_{hashlib.md5().hexdigest()}"
E       ValueError: [digital envelope routines] disabled for fips
```
https://github.com/pandas-dev/pandas/blob/main/pandas/core/reshape/merge.py#L1295-L1314

The only workaround is to disable FIPS mode, which is insecure. 

### Expected Behavior

Pandas should be able to perform a cross merge on systems operating in FIPS mode. md5 is a forbidden algorithm in FIPS environments, so the easiest method is to change the md5 hash to a sha256 (or greater)

### Installed Versions

<details>

 python -c 'import pandas as pd; pd.show_versions()'

INSTALLED VERSIONS
------------------
commit           : e8093ba372f9adfe79439d90fe74b0b5b6dea9d6
python           : 3.8.13.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.104-linuxkit
Version          : #1 SMP Wed Mar 9 19:05:23 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.3
numpy            : 1.23.1
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 63.4.2
pip              : 22.2.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.3
jinja2           : 3.1.2
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : 2022.7.1
gcsfs            : None
markupsafe       : 2.1.1
matplotlib       : 3.5.2
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 7.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : 1.4.40
tables           : None
tabulate         : 0.8.10
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #48024 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @mroeschke Any idea how to test this? Did not add it for the test, do you think thats necessary?

### CODE REVIEW

File: pandas/core/reshape/merge.py [hunk 1]
```diff
@@ -1311,7 +1313,16 @@ def _create_cross_configuration(
...
+        _md5: Callable
+        if PY39:
+
+            def _md5(_hashlib_md5=hashlib.md5):
+                return _hashlib_md5(usedforsecurity=False)
+
+        else:
+            _md5 = hashlib.md5
+
+        cross_col = f"_cross_{_md5().hexdigest()}"
```
Comment (mroeschke) [inline pandas/core/reshape/merge.py:1325]: Curious, why was md5 chosen originally? (Looks like you were the original author https://github.com/pandas-dev/pandas/pull/37864 ;))

If the intention was to generate a unique column for joining later, could we just use `uuid.uuid4()` instead?

Comment (phofl) [reply pandas/core/reshape/merge.py:1325]: Yes, the intention was generating a unique column, not sure anymore why md5 explicitly. Can change I think

**Comment by mroeschke** [comment]
@szelenka, could you confirm using uuid4 would be a sufficient replacement for this use case (if you can test locally)?

From what I am seeing online, I don't see a really a convenient way of testing this in our CI system

**Comment by szelenka** [comment]
@mroeschke yes this seems to work okay:
```
>>> ssl.FIPS_mode_set(0)
>>> ssl.FIPS_mode_set(1)

# Old method
>>> cross_col = f"_cross_{hashlib.md5().hexdigest()}"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: [digital envelope routines] disabled for fips

# New method
>>> cross_col = f"_cross_{uuid.uuid4()}"
>>> cross_col
'_cross_0a42418d-46db-40c9-8cbf-72d672d9863a'
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl and @szelenka for confirming.

**Pull request merged**